Adding bulk price settings view controller

chrisKns opened the pull request:
Part of: #6242 & #6239

### Description
Adds the view controller that allows the user to enter a new regular price and bulk update all the product variations.
This view controller is pushed when the regular price cell is tapped in the bulk update screen #6239.
The view controller has a tableview with a single cell `UnitInputTableViewCell` that contains the textfield where the user enters the price. The tableView does not extent to cover the whole view, an area at the bottom of the view is the place of the save button(of type `ButtonActivityIndicator`). The keyboard is monitored using the `KeyboardFrameObserver` and the constraint of the save button to the bottom is adjusted in order to always be 16 points above the keyboard.

The `UnitInputTableViewCell` currently has the same style as in the single variation price update, I will provide the style to match the mockups in a followup pull request(that has only the price input on the left, we have a sub-task for this).
Price change events are send to the view model, and it updates the state of the button accordingly. The view controller receives events for the state of the button and is enabled if we have a non empty price.

When the save button is tapped the API call is triggered, the keyboard is dismissed and for the duration of the call the button displays an activity indicator(and can not be tapped).
If the bulk update is successful the view controller is popped from the navigation stack. In case of error the user will remain in the current screen. In both cases an `Notice` will be displayed that will be added in a follow up pull request(, we have a sub-task for this).  
The view controller also has a footer text that displays a text informing the user about the current price of the variations and the number of variations to be updated. Plural variations have been taken in to account. To create this text we are passing in the initializer of the `BulkUpdatePriceSettingsViewModel` a  BulkUpdateOptionsModel instead of an array of product variations.

### Testing instructions
1. Goto the products tab
2. Select a product with variations
3. Go to the list of variations
4. Tap on the "more" button in the navigation bar
5. Select the "Bulk update"
6. The bulk update screen should appear, select the regular price option
7. The price setting screen should appear
8. Enter a valid price, the save button should be enabled only if the price is non empty
9. Bellow the price field a text should be displayed that depending if the variations have the same price or not the footer text have the following values:
Current price is <formated price>. The price will be updated for <number of variations> variation[s].
Current prices are mixed.. The price will be updated for <number of variations> variation[s].
Current price is not set. The price will be updated for <number of variations> variation[s].
10. Tap save button, the keyboard should be dismissed and a loading indicator should be displayed in the save button
11. The price should be updated and the user returned to the previous bulk update screen




https://user-images.githubusercontent.com/96764631/157611395-15e18cf0-defc-4108-8e0c-b7724a86af07.mp4



https://user-images.githubusercontent.com/96764631/157609296-a1ccfea3-cd5f-4820-9d39-c53adf6650bb.mp4



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

csknns committed:
Adding bulk price settings view controller

chrisKns commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Bulk Edit Price/BulkUpdatePriceSettingsViewModel.swift:
```diff
@@ -140,4 +155,57 @@ final class BulkUpdatePriceSettingsViewModel {
+
+        switch bulkUpdateOptionsModel.bulkValueOf(editingPriceType.keyPathForPriceType()) {
+        case .none:
+            return [Localization.currentPriceNoneFooter, numberOfVariationsText].joined(separator: " ")
```
Since they are complete & separate sentences I decided to concatenate them.

csknns committed:
Fixed comment

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Bulk Edit Price/BulkUpdatePriceSettingsViewModel.swift:
```diff
@@ -106,27 +117,31 @@ final class BulkUpdatePriceSettingsViewModel {
     private func updateButtonStateBasedOnCurrentPrice() {
-        guard let price = currentPrice, price.isNotEmpty else {
-            saveButtonState = .disabled
+        // While the keyboard is in loading state do not change the state
```
Did you mean button?
```suggestion
        // While the action button is a loading state do not change the state
```

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Bulk Edit Price/BulkUpdatePriceSettingsViewModel.swift:
```diff
@@ -140,4 +155,57 @@ final class BulkUpdatePriceSettingsViewModel {
+
+    /// Returns the footer text to be displayed with information about the current bulk price and how many variations will be updated.
+    ///
+    private func footerText() -> String {
+        let numberOfVariations = bulkUpdateOptionsModel.productVariations.count
+        let numberOfVariationsText = String.pluralize(numberOfVariations,
+                                                      singular: Localization.variationsNumberSingularFooter,
+                                                      plural: Localization.variationsNumberPlurarFooter)
+
+        switch bulkUpdateOptionsModel.bulkValueOf(editingPriceType.keyPathForPriceType()) {
+        case .none:
+            return [Localization.currentPriceNoneFooter, numberOfVariationsText].joined(separator: " ")
+        case .mixed:
+            return [Localization.currentPriceMixedFooter, numberOfVariationsText].joined(separator: " ")
+        case let .value(price):
+            let currentPriceText = String.localizedStringWithFormat(Localization.currentPriceFooter, formatPriceString(price))
+            return [currentPriceText, numberOfVariationsText].joined(separator: " ")
+        }
+    }
```
To not calculate this twice(Since it was calculated in the previews VC) I wonder if we should just provide the price type as a parameter 🤔

chrisKns replied:
We could but we are not storing the previous calculated result, so we would need to store it in the `BulkUpdateViewModel.swift` and updated it then the `BulkUpdateOptionsModel` changes (probably with a property observer), instead of doing this for all values(regular, sale price, dates etc.) do you think it would be better to simply cache the results in a opaque manner in `BulkUpdateOptionsModel` ?

Ecarrion replied:
Nah, that is ok, no need to overcomplicate it.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Bulk Edit Price/BulkUpdatePriceViewController.swift:
```diff
@@ -0,0 +1,193 @@
+        keyboardFrameObserver.startObservingKeyboardFrame(sendInitialEvent: true)
+    }
+
+    /// Configures the  title and appearance of the save button
```
```suggestion
    /// Configures the title and appearance of the save button
```

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Bulk Edit Price/BulkUpdatePriceViewController.swift:
```diff
@@ -0,0 +1,193 @@
+            if state == .loading {
+                self?.saveButton.showActivityIndicator()
+                // Dismiss the keyboard while we are in loading state
+                self?.view.endEditing(true)
```
This instruction does not seem right here. Don't you think we should move it to `saveButtonTapped`?
Or maybe expose a different state for the keyboard handling 🤔

chrisKns replied:
I think It will be easier to follow & simpler if I added to the `saveButtonTapped()`, we will also have the extra space to display the validation errors if any (in landscape it left a little cramped). I updated the PR.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Bulk Edit Price/BulkUpdatePriceViewController.swift:
```diff
@@ -0,0 +1,193 @@
+    func configure(_ cell: UITableViewCell, at indexPath: IndexPath) {
+        switch cell {
+        case let cell as UnitInputTableViewCell:
+            let cellViewModel = Product.createRegularPriceViewModel(regularPrice: viewModel.currentPrice,
+                                                                    using: ServiceLocator.currencySettings) { [weak self] value in
+                self?.viewModel.handlePriceChange(value)
+            }
```
I know this is not your code but it does not feel right for a `Product` to create a view model. 😅

chrisKns replied:
Yes I know, it felt out of place, should I open an improvement issue for this? I think an extension on the view model should be better.

I have a followup PR that adds the proper style (to match the mockups with the input on the left) so I can improve this by moving it to an extension in the view model.

Ecarrion replied:
Feel free to open the issue :-)

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Variations/Bulk Update/BulkUpdateViewController.swift:
```diff
@@ -157,6 +157,16 @@ final class BulkUpdateViewController: UIViewController {
+    ///
+    private func navigateToEditPriceSettings() {
+        let bulkUpdatePriceSettingsViewModel = viewModel.viewModelForBulkUpdatePriceOfType(.regular, priceUpdateDidFinish: { [weak self] in
+            self?.navigationController?.popToRootViewController(animated: true)
```
Why poping to `root`? Shouldn't it just pop to the previous one? 
`BulkUpdateViewController.swift` does not know who presented it to what else is in the stack, so I guess is not safe to make the assumption that it will always be the root.

chrisKns replied:
Indeed! popping to the view controller that did the push is safer here. I pushed a fix.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Variations/Bulk Update/BulkUpdateViewController.swift:
```diff
@@ -157,6 +157,16 @@ final class BulkUpdateViewController: UIViewController {
+            self?.navigationController?.popToRootViewController(animated: true)
+        })
+        let viewController = BulkUpdatePriceViewController(viewModel: bulkUpdatePriceSettingsViewModel)
+        navigationController?.pushViewController(viewController, animated: true)
```
```suggestion
        show(viewController, sender: nil)
```

Ecarrion reviewed (commented):
The PR looks good in general! Although there are a lot of changes.

I noticed some odd behaviors but I think they would be more clear once we add error notices.

Also spotted two issues, one related to the PR and one that is not. Feel free to create separate issues for them.

1. When dealing with accessibility fonts, the table and the button overlaps.

<img width="443" alt="accessbility" src="https://user-images.githubusercontent.com/562080/157744042-09284db4-be0c-43d1-875b-a857bdb69469.png">

2. It looks like the first time when opening a product variation lost, the `...` button is not available

https://user-images.githubusercontent.com/562080/157744212-24a27402-e9f3-4678-b4a4-10ed53cc4519.mov

chrisKns committed:
Update WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Bulk Edit Price/BulkUpdatePriceSettingsViewModel.swift

Co-authored-by: Ernesto Carrión <carrion.silver@gmail.com>

chrisKns committed:
Update WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Bulk Edit Price/BulkUpdatePriceViewController.swift

Co-authored-by: Ernesto Carrión <carrion.silver@gmail.com>

csknns committed:
Popping to the view contoller that did the push instead of the root

chrisKns committed:
Update WooCommerce/Classes/ViewRelated/Products/Variations/Bulk Update/BulkUpdateViewController.swift

Simplified code pushing on navigation stack

Co-authored-by: Ernesto Carrión <carrion.silver@gmail.com>

csknns committed:
Moved keyboard dismissal after the save button is tapped

csknns committed:
Fixed issue with tableView content overlaping button in disable state

chrisKns commented:
> The PR looks good in general! Although there are a lot of changes.

🎉 I think I should have made a separate PR for the footer information so make it easier to review, I make them more focused in the future!

> I noticed some odd behaviors but I think they would be more clear once we add error notices.
> 
> Also spotted two issues, one related to the PR and one that is not. Feel free to create separate issues for them.
> 
> 1. When dealing with accessibility fonts, the table and the button overlaps.
> 

This seems to be caused by the transparency of the button in the disabled state. I added a background color in 1a3171825 to match the tableView and it was fixed. I also adjusted the content inset of the tableView when the keyboard appears to be able to scroll though the tableView's content if the keyboard hides it.
 
> 
> 2. It looks like the first time when opening a product variation lost, the `...` button is not available
> 

This is weird, I also started to notice this, opened #6414 and added it to the parent's issue [task list](https://github.com/woocommerce/woocommerce-ios/issues/5762#issuecomment-1043347770)

Ecarrion reviewed (approved):
Looks good! Nice to see the feature near its completion!
